recipes/openai_ab.py

```diff
@@ -0,0 +1,365 @@
+        msg.good(f"You preferred {pref}")
+        rows = [(name, count) for name, count in counts.most_common()]
+        msg.table(rows, aligns=("l", "r"))
+
+    def _get_response_batch(self, inputs: List[PromptInput]) -> List[Dict[str, str]]:
+        name1, name2 = self._choose_rivals()
+        prompts = []
+        for input_ in inputs:
+            prompts.append(self._get_prompt(name1, input_.prompt_args))
+            prompts.append(self._get_prompt(name2, input_.prompt_args))
+        responses = self._get_responses(prompts)
+        assert len(responses) == len(inputs) * 2
+        output = []
+        # Pair out the responses. There's a fancy
+        # zip way to do this but I think that's less
+        # readable
+        for i in range(0, len(responses), 2):
+            output.append({name1: responses[i], name2: responses[i + 1]})
+        return output
+
+    def _choose_rivals(self) -> Tuple[str, str]:
+        assert len(self.prompts) == 2
+        return tuple(sorted(self.prompts.keys()))
+
+    def _get_prompt(self, name: str, args: Dict) -> str:
+        return self.prompts[name].render(**args)
+
+    def _get_responses(self, prompts: List[str]) -> List[str]:
+        headers = {
+            "Authorization": f"Bearer {self.openai_api_key}",
+            "OpenAI-Organization": self.openai_api_org,
+            "Content-Type": "application/json",
+        }
+        r = _retry429(
+            lambda: httpx.post(
+                "https://api.openai.com/v1/completions",
+                headers=headers,
+                json={
+                    "model": self.model,
+                    "prompt": prompts,
+                    "temperature": self.openai_temperature,
+                    "max_tokens": self.openai_max_tokens,
+                },
+                timeout=self.openai_timeout_s,
+            ),
+            n=self.openai_n,
+            timeout_s=self.openai_timeout_s,
+        )
+        r.raise_for_status()
+        responses = r.json()
+        return [responses["choices"][i]["text"] for i in range(len(prompts))]
+
+    def _make_example(
+        self, id: str, input: str, responses: Dict[str, str], randomize: bool
+    ) -> Dict:
+
+        question = {
+            "id": id,
+            "text": input,
+            "options": [],
+        }
+        response_pairs = list(responses.items())
+        if randomize:
+            random.shuffle(response_pairs)
+        else:
+            response_pairs = list(sorted(response_pairs))
+        for name, value in response_pairs:
+            question["options"].append({"id": name, "text": value})
+        return question
+
+
+@prodigy.recipe(
+    "ab.openai.prompts",
+    dataset=("Dataset to save answers to", "positional", None, str),
+    inputs_path=("Path to jsonl inputs", "positional", None, Path),
+    display_template_path=(
+        "Template for summarizing the arguments",
+        "positional",
+        None,
+        Path,
+    ),
```

Comment:
I worry that this parameter will be confusing to users. Using it as a descriptive title means it'll always be pretty similar to one of the actual prompts, adding to the confusion.
I wonder whether instead we should have a more structured title on the Prodigy UI? So instead of 
```
Poem about just vibing out in the style of John Donne
```
it would read something more "technical", showing the actual slots literally, like 
```
Topic: "just vibing out"
Author: "John Donne"
```

Reply:
It could just be passed as `"meta"` information on the example. I'll have a look!

Reply:
I just expanded `._make_example` to also include the meta information, which looks like this in the annotation interface: 

![CleanShot 2023-01-26 at 13 52 07](https://user-images.githubusercontent.com/1019791/214840122-50f4348f-1815-438e-9ed6-975500fc742f.png)

![Uploading CleanShot 2023-01-26 at 13.52.23.png…]()

Reply:
Apon reflection, I think I like it better with the header around. 

![CleanShot 2023-01-26 at 14 02 27](https://user-images.githubusercontent.com/1019791/214842271-9ca9dbee-5329-469e-8082-8949d1030ba7.png)

It might still be nice to have it optional, but I'm not super bullish on that. I'll keep Matt's implementation here.